### PR TITLE
feat: common.attachAutoResizer

### PR DESCRIPTION
**Pull request by iwoplaza**
This PR adds a new common helper called `attachAutoResizer` which adapts the canvas' internal resolution to match its physical pixel size, with additional safe-guards. It's a web only helper, not to be used in React Native apps.

The examples were updated to use this helper, which replaces the need for the example environment to handle the canvas resizing. This also makes the examples much more portable, as resizing a canvas can be tricky to get right, and now it's solved in every example.

### CODE REVIEW

File: packages/typegpu/src/common/attachAutoResizer.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    }
+  });
+
+  try {
+    observer.observe(canvas, { box: 'device-pixel-content-box' });
+  } catch {
+    observer.observe(canvas, { box: 'content-box' });
+  }
+
+  return () => {
```
Comment (deluksic) [inline packages/typegpu/src/common/attachAutoResizer.ts]: What do you think about returning a `{ detach: () => {} }` so its self-documenting?

Comment (deluksic) [reply packages/typegpu/src/common/attachAutoResizer.ts]: then the rest of the code will read `autoResizer.detach()`

File: packages/typegpu/src/common/attachAutoResizer.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import type { TgpuRoot } from '../core/root/rootTypes.ts';
+
+export interface AttachAutoResizerOptions {
+  root: TgpuRoot;
+  canvas: HTMLCanvasElement;
+  onResize?(): void;
```
Comment (deluksic) [inline packages/typegpu/src/common/attachAutoResizer.ts]: `onResize` will usually want to do something with the new canvas size, so might as well pass it in

Comment (iwoplaza) [reply packages/typegpu/src/common/attachAutoResizer.ts]: I think devs usually have access to the canvas they passed in, and they can access `canvas.width` and `canvas.height`. That way we can avoid an additional array allocation on each resize.

File: packages/typegpu/src/common/attachAutoResizer.ts (hunk 1 above)
Comment (iwoplaza) [reply packages/typegpu/src/common/attachAutoResizer.ts]: Great idea, changed it 🫡

**Comment by iwoplaza** [comment]
@pullfrog Review this one please 🙏

File: packages/typegpu/src/common/attachAutoResizer.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import type { TgpuRoot } from '../core/root/rootTypes.ts';
+
+export interface AttachAutoResizerOptions {
+  root: TgpuRoot;
+  canvas: HTMLCanvasElement;
+  onResize?: (() => void) | undefined;
```
[the remark this answers is not in the document]
Comment (deluksic) [reply packages/typegpu/src/common/attachAutoResizer.ts:6]: This is not true. In codebases with `exactOptionalPropertyTypes`, having `| undefined` has meaning. If you omit `undefined` you are explicitly saying `{ onResize: undefined }` can't be sent in.

File: apps/typegpu-docs/src/examples/rendering/xor-dev-runner/index.html [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<canvas data-fit-to-container></canvas>
+<canvas class="w-full h-full" />
 <p class="absolute px-2 py-1 rounded-xl top-2 mx-auto text-lg text-center text-white bg-black/50">
```
[the remark this answers is not in the document]
Comment (aleksanderkatan) [reply apps/typegpu-docs/src/examples/rendering/xor-dev-runner/index.html:2]: From what I google, the clanker is right, and this is equivalent to just not closing the canvas element at all. This appears in multiple places, try regex replacing `<canvas(.*)/>` with `<canvas$1></canvas>`

File: apps/typegpu-docs/src/examples/algorithms/bitonic-sort/index.html [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<canvas></canvas>
+<canvas class="w-full h-full" />
```
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/algorithms/bitonic-sort/index.html:1]: Is this what we want?
Image

File: apps/typegpu-docs/src/examples/tests/uniformity/index.ts [hunk 6]
```diff
@@ -83,14 +83,17 @@ export const controls = defineControls({
   },
 });
 
-const resizeObserver = new ResizeObserver(() => {
-  canvasRatioUniform.write(canvas.width / canvas.height);
-  redraw();
+const autoResizer = common.attachAutoResizer({
```
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/tests/uniformity/index.ts:86]: Were three examples intentionally excluded from this change?

File: apps/typegpu-docs/src/examples/tests/texture-test/index.ts
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/tests/texture-test/index.ts:1]: This example didn't use to stretch

File: apps/typegpu-docs/src/examples/tests/sdf-docs-examples/index.ts
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/tests/sdf-docs-examples/index.ts:1]: This example didn't use to stretch

File: apps/typegpu-docs/src/examples/tests/rc-docs-examples/index.ts
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/tests/rc-docs-examples/index.ts:1]: This example didn't use to stretch

File: apps/typegpu-docs/src/examples/tests/log-test/index.ts [hunk 7]
```diff
@@ -288,7 +288,10 @@ export const controls = defineControls({
   },
 });
 
+const autoResizer = common.attachAutoResizer({ root, canvas });
```
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/tests/log-test/index.ts:291]: This example shouldn't have a resizer.

File: apps/typegpu-docs/src/examples/algorithms/mnist-inference/index.html
Comment (aleksanderkatan) [inline apps/typegpu-docs/src/examples/algorithms/mnist-inference/index.html:1]: This example is now broken

**Review by aleksanderkatan — CHANGES_REQUESTED** [review]
Here are some things I noticed:

Many examples (mostly orbit camera ones, but not all of those) throw when resizing the page to be so small that only example list is displayed.
http://localhost:4321/TypeGPU/examples#example=rendering--os-awards a FiniteMathAssumption error is thrown (before, it had wgsl errors), it seems that it works regardless
http://localhost:4321/TypeGPU/examples#example=simulation--gravity -||-
http://localhost:4321/TypeGPU/examples#example=rendering--pom -||-
http://localhost:4321/TypeGPU/examples#example=rendering--xor-dev-runner -||-
http://localhost:4321/TypeGPU/examples#example=rendering--function-visualizer -||-
http://localhost:4321/TypeGPU/examples#example=rendering--phong-reflection -||-
http://localhost:4321/TypeGPU/examples#example=rendering--render-bundles -||-
http://localhost:4321/TypeGPU/examples#example=rendering--render-bundles-with -||-
http://localhost:4321/TypeGPU/examples#example=simple--ripple-cube -||-, also has many WGSL warnings
http://localhost:4321/TypeGPU/examples#example=simple--mesh-skinning crashes

http://localhost:4321/TypeGPU/examples#example=algorithms--jump-flood-distance started flashing black when resizing
